### DISCUSSION TITLE
MidiFile Pause, Resume, Stop?, Loop, Beats, Measures

### DISCUSSION THREAD

**Discussion by gmeader** [mido/mido#369] [Q&A]
I'm working on a drum machine that works by playing MIDI files. I'm kind of a newbie, so I have a bunch of questions.
The file playback needs to keep looping, unless paused or stopped.
I need to be able to **pause** and **resume** playing from where the pause occurred and that along with **stop** needs to happen on a beat/measure boundary. 

1. is mido the right library to use for this? Should I try implementing some of this and do a PR?
2. I tried implementing pause/resume in the playmidi example by waiting in the play() loop, but when I, resume, it plays a whole bunch of notes quickly to catch up (the clock in mido play() does not pause) So I think Pause and Resume might need to be implemented within the library?
3. How do I loop so that the beat matches when it loops?
4. Can I convert the time in seconds returned by the yield in play() to beats/measures? I guess I can convert seconds back to ticks and then use PPQ to calculate the beat and time-signature to calculate the measure. Will the conversion from seconds to ticks for this be accurate?

**Comment by rdoursenaud** [comment]
Hi, welcome and good luck on your project!

1. Yes! MIDO seems to be a great fit if you don't want to manipulate raw MIDI data. However, it may need some love before it can provide every service needed. Of course, some parts obviously reside inside your application and not the library and decisions will have to be made where each responsibilities starts and ends ;)
PRs are always a good thing even if I haven't personally seen the maintainer in a while. Hope he's doing well and will be coming back soon.

2. I'm not familiar with this side of the lib but a quick glance at the code reveals that it could indeed benefit from `pause()`, `resume()` and `loop()` methods! I'd happily review your code.

3. Tricky subject. We'll have to figure this out. I guess that we'll need to have an internal Beats, Bars & Ticks (BBT) representation of the time to make this happen. Or it could be the added value in your app ;)

4. Not very familiar with this side either (yet!) but I know there is at least somewhat relevant opened Issues (#102, #351, #359) and a PR (#114) that you may want to integrate locally until its merged.
